{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notes\n",
    "\n",
    "- Data taken from here: https://github.com/alibashir/aav (File: allseqs_20191230.csv.zip)\n",
    "- Just sequence (AA) data exists as well here: https://www.dropbox.com/sh/lmd8wmgibz24r2h/AADmgvc_0Q5mylwgGf97cTC-a/ML_and_Analysis/model_and_dataset_analysis/allseqs_20191230.csv.zip?dl=0\n",
    "- [This notebook](https://github.com/churchlab/Deep_diversification_AAV/blob/main/Parsing_pipeline/Step3_compute_selection_scores.ipynb) suggest that the \"target\" column is `GAS1_virus_S` in the CSV file\n",
    "- [This repository](https://github.com/google-research/google-research/tree/master/aav/model_training) suggest that the wild type AA sequence is 28 residues and is `DEEEIRTTNPVATEQYGSVSTNLQRGNR`\n",
    "- Cutoff via gaussian mixture copied from: https://colab.research.google.com/github/google-research/google-research/blob/master/aav/model_and_dataset_analysis/data_prep.ipynb\n",
    "- The UniProt sequence is here: https://www.uniprot.org/uniprot/P03135\n",
    "- In UniProt notation (sequence starts at 1), the region on P03135 where the sequence mutates is `[561, 588]`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (1.78)\n",
      "Requirement already satisfied: numpy in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from biopython) (1.20.2)\n",
      "Requirement already satisfied: pandas in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (1.2.4)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from pandas) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.3 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from pandas) (2021.1)\n",
      "Requirement already satisfied: numpy>=1.16.5 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from pandas) (1.20.2)\n",
      "Requirement already satisfied: six>=1.5 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from python-dateutil>=2.7.3->pandas) (1.16.0)\n",
      "Requirement already satisfied: matplotlib in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (3.4.2)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (2.4.7)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (0.10.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (2.8.1)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (8.2.0)\n",
      "Requirement already satisfied: numpy>=1.16 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (1.20.2)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (1.3.1)\n",
      "Requirement already satisfied: six in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from cycler>=0.10->matplotlib) (1.16.0)\n",
      "Requirement already satisfied: scikit-learn in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (0.24.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from scikit-learn) (2.1.0)\n",
      "Requirement already satisfied: scipy>=0.19.1 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from scikit-learn) (1.6.3)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from scikit-learn) (1.20.2)\n",
      "Requirement already satisfied: joblib>=0.11 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from scikit-learn) (1.0.1)\n",
      "Requirement already satisfied: python-Levenshtein in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (0.12.2)\n",
      "Requirement already satisfied: setuptools in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from python-Levenshtein) (49.6.0.post20210108)\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython\n",
    "!pip install pandas\n",
    "!pip install matplotlib\n",
    "!pip install scikit-learn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import math\n",
    "import random\n",
    "import numpy\n",
    "\n",
    "from pathlib import Path\n",
    "from pandas import read_csv, DataFrame\n",
    "\n",
    "from helpers import read_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "pandas.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = Path('') / '..' / 'data' / 'aav'\n",
    "aav_data_path = data_path / 'allseqs_20191230.csv'\n",
    "\n",
    "split_path = Path('') / '..' / 'splits' / 'aav'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "aav_data = read_csv(aav_data_path)\n",
    "\n",
    "P03135 = read_fasta(data_path / 'P03135.fasta')[0]\n",
    "region = (561,588)\n",
    "reference_region = \"DEEEIRTTNPVATEQYGSVSTNLQRGNR\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sequence', 'partition', 'mutation_sequence', 'num_mutations',\n",
       "       'num_edits', 'viral_selection', 'is_viable'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aav_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "designed\n",
      "rand\n",
      "single\n",
      "stop\n",
      "cnn_designed_plus_rand_train_walked\n",
      "lr_rand_doubles_plus_single_walked\n",
      "cnn_rand_doubles_plus_single_seed\n",
      "cnn_standard_walked\n",
      "rnn_rand_doubles_plus_singles_seed\n",
      "lr_standard_walked\n",
      "lr_rand_doubles_plus_single_seed\n",
      "lr_designed_plus_rand_train_seed\n",
      "cnn_rand_doubles_plus_single_walked\n",
      "previous_chip_nonviable\n",
      "cnn_designed_plus_rand_train_seed\n",
      "lr_standard_seed\n",
      "rnn_designed_plus_rand_train_walked\n",
      "lr_designed_plus_rand_train_walked\n",
      "random_doubles\n",
      "rnn_standard_seed\n",
      "previous_chip_viable\n",
      "rnn_standard_walked\n",
      "rnn_designed_plus_rand_train_seed\n",
      "cnn_standard_seed\n",
      "rnn_rand_doubles_plus_singles_walked\n",
      "singles\n",
      "wild_type\n"
     ]
    }
   ],
   "source": [
    "for partition in aav_data['partition'].unique():\n",
    "    print(partition)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the wild type sequence from the data\n",
    "\n",
    "wt = aav_data[aav_data['partition'] == \"wild_type\"].to_dict(orient='records')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure that all assumptions on the wild type are correct\n",
    "\n",
    "assert(wt['sequence'] == reference_region)\n",
    "\n",
    "# Important: region[0] -1 because UniProt notation! (0 in Python == 1 in UniProt)\n",
    "assert(wt['sequence'] == str(P03135[region[0]-1 : region[1]].seq))\n",
    "\n",
    "# Make sure chopping the sequence works as expected:\n",
    "\n",
    "assert(\n",
    "    str(P03135.seq) ==\n",
    "    str(P03135[:region[0]-1].seq) + str(P03135[region[0]-1 : region[1]].seq) + str(P03135[region[1]:].seq)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove data where scores are +/- inf\n",
    "\n",
    "aav_data.replace([math.inf, -math.inf], math.nan, inplace=True)\n",
    "aav_data.dropna(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We need to aggregate synonymous variants if operating on AA sequences\n",
    "# Take the score and devide by number of synonymous WT AA seuqneces!\n",
    "\n",
    "def aggregate_rows(dataframe_slice):\n",
    "    \n",
    "    first_element = dataframe_slice.iloc[0]\n",
    "    \n",
    "    return {\n",
    "        'mutation_mask': first_element.mutation_sequence,\n",
    "        'mutated_region': first_element.sequence,\n",
    "        'category': first_element.partition,\n",
    "        'number_of_mutations': first_element.num_mutations,\n",
    "        'levenshtein_distance': first_element.num_edits,\n",
    "        'binary_score': first_element.is_viable,\n",
    "        'score': dataframe_slice.viral_selection.mean()\n",
    "    }\n",
    "\n",
    "grouped_protein_data = aav_data.groupby(['sequence', 'partition']).apply(aggregate_rows).values\n",
    "\n",
    "protein_data = DataFrame.from_records(grouped_protein_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def region_cut(new_sequence):\n",
    "    filtered_sequence = new_sequence.upper().replace('*','')\n",
    "    return str(P03135[:region[0]-1].seq) + filtered_sequence + str(P03135[region[1]:].seq)\n",
    "\n",
    "protein_data['full_aa_sequence'] = protein_data['mutated_region'].apply(region_cut)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the the manuscript:\n",
    " \n",
    "> To avoid overfitting to experimental noise, rather than predicting the quantitative production efficiency we used binary classification models to predict whether each sequence variant is viable (Supplementary Fig. 1), as defined by a threshold fit to best separate positive and negative controls (WT replicas, and variants containing stop codons, respectively).\n",
    "\n",
    "Apparently, by plotting `score` one should see a binomial. The cutoff is set somewhere beteen -2.5 and 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYMAAAD4CAYAAAAO9oqkAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAX3ElEQVR4nO3df4xd5Z3f8fdnYUlYGoIJ2SmyUY0UKyuCFQoj8CqrajZsjCHVmq0SBEKLSd14pcA2W1nqmvYPtyFIpGpKQ5WgusXFRLtxKF2EFZx4XSdXUf8wARIKMQQxS8zaFj92Y350EiXRpN/+cZ9p7jozvnd+eO4MvF/SaM75nuec89zHd/yZc+6Zc1JVSJLe3n5t2B2QJA2fYSBJMgwkSYaBJAnDQJIEnD7sDszVeeedV6tXrx7a/n/84x9z1llnDW3/y4Xj1J9j1J9jNJh+4/TEE0/8bVW9d7plyzYMVq9ezeOPPz60/Xc6HcbGxoa2/+XCcerPMerPMRpMv3FK8uJMyzxNJEkyDCRJhoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIklvFfIEtvRau3PTLndQ/f+dEF7InebgY6MkjyL5IcSvL9JF9J8s4kFyZ5NMl4kq8mOaO1fUebH2/LV/ds57ZWfy7JVT31Da02nmTbgr9KSdJJ9Q2DJCuBfw6MVtXFwGnA9cDngLuq6n3Aa8Dmtspm4LVWv6u1I8lFbb0PABuALyU5LclpwBeBq4GLgBtaW0nSIhn0M4PTgTOTnA78BvAS8GHgwbZ8F3Btm97Y5mnLr0ySVt9dVT+rqh8C48Dl7Wu8ql6oqp8Du1tbSdIi6RsGVXUM+PfAX9MNgTeAJ4DXq2qyNTsKrGzTK4Ejbd3J1v49vfUT1pmpLklaJH0/QE6ygu5v6hcCrwP/ne5pnkWXZAuwBWBkZIROpzOMbgAwMTEx1P0vF45Tf71jtHXt5Mkbn8RbeZx9Hw1mPuM0yNVEvwf8sKr+BiDJXwAfAs5Jcnr77X8VcKy1PwZcABxtp5XeDfyopz6ld52Z6n9HVe0AdgCMjo7WMO9v7v3VB+M49dc7RjfP52qiG8cWpkNLkO+jwcxnnAb5zOCvgXVJfqOd+78SeAb4FvCx1mYT8HCb3tPmacu/WVXV6te3q40uBNYA3wEeA9a0q5POoPsh8545vRpJ0pz0PTKoqkeTPAh8F5gEvkf3t/NHgN1JPttq97ZV7gW+nGQcOE73P3eq6lCSB+gGySRwS1X9AiDJrcA+ulcq7ayqQwv3EiVJ/Qz0R2dVtR3YfkL5BbpXAp3Y9qfAx2fYzh3AHdPU9wJ7B+mLJGnheTsKSZJhIEkyDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkMUAYJHl/kid7vt5M8idJzk2yP8nz7fuK1j5J7k4ynuSpJJf2bGtTa/98kk099cuSPN3Wubs9a1mStEj6hkFVPVdVl1TVJcBlwE+Ah4BtwIGqWgMcaPMAV9N92P0aYAtwD0CSc+k+OvMKuo/L3D4VIK3NJ3vW27AQL06SNJjZnia6EvirqnoR2AjsavVdwLVteiNwf3UdBM5Jcj5wFbC/qo5X1WvAfmBDW3Z2VR2sqgLu79mWJGkRnD7L9tcDX2nTI1X1Upt+GRhp0yuBIz3rHG21k9WPTlP/FUm20D3aYGRkhE6nM8vuL5yJiYmh7n+5cJz66x2jrWsn57ydt/I4+z4azHzGaeAwSHIG8PvAbScuq6pKUnPqwSxU1Q5gB8Do6GiNjY2d6l3OqNPpMMz9LxeOU3+9Y3TztkfmvJ3DN44tTIeWIN9Hg5nPOM3mNNHVwHer6pU2/0o7xUP7/mqrHwMu6FlvVaudrL5qmrokaZHMJgxu4JeniAD2AFNXBG0CHu6p39SuKloHvNFOJ+0D1idZ0T44Xg/sa8veTLKuXUV0U8+2JEmLYKDTREnOAj4C/FFP+U7ggSSbgReB61p9L3ANME73yqNPAFTV8SS3A4+1dp+pquNt+lPAfcCZwNfblyRpkQwUBlX1Y+A9J9R+RPfqohPbFnDLDNvZCeycpv44cPEgfZEkLTz/AlmSZBhIkgwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkYRhIkhgwDJKck+TBJD9I8myS305ybpL9SZ5v31e0tklyd5LxJE8lubRnO5ta++eTbOqpX5bk6bbO3e3xl5KkRTLokcEXgG9U1W8BHwSeBbYBB6pqDXCgzQNcDaxpX1uAewCSnAtsB64ALge2TwVIa/PJnvU2zO9lSZJmo28YJHk38I+AewGq6udV9TqwEdjVmu0Crm3TG4H7q+sgcE6S84GrgP1VdbyqXgP2AxvasrOr6mB7ZOb9PduSJC2CQZ6BfCHwN8B/S/JB4Ang08BIVb3U2rwMjLTplcCRnvWPttrJ6kenqf+KJFvoHm0wMjJCp9MZoPunxsTExFD3v1w4Tv31jtHWtZNz3s5beZx9Hw1mPuM0SBicDlwK/HFVPZrkC/zylBAAVVVJak49mIWq2gHsABgdHa2xsbFTvcsZdTodhrn/5cJx6q93jG7e9sict3P4xrGF6dAS5PtoMPMZp0E+MzgKHK2qR9v8g3TD4ZV2iof2/dW2/BhwQc/6q1rtZPVV09QlSYukbxhU1cvAkSTvb6UrgWeAPcDUFUGbgIfb9B7gpnZV0TrgjXY6aR+wPsmK9sHxemBfW/ZmknXtKqKberYlSVoEg5wmAvhj4M+SnAG8AHyCbpA8kGQz8CJwXWu7F7gGGAd+0tpSVceT3A481tp9pqqOt+lPAfcBZwJfb1+SpEUyUBhU1ZPA6DSLrpymbQG3zLCdncDOaeqPAxcP0hdJ0sLzL5AlSYaBJMkwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkDANJEoaBJIkBwyDJ4SRPJ3kyyeOtdm6S/Umeb99XtHqS3J1kPMlTSS7t2c6m1v75JJt66pe17Y+3dbPQL1SSNLPZHBn8blVdUlVTTzzbBhyoqjXAgTYPcDWwpn1tAe6BbngA24ErgMuB7VMB0tp8sme9DXN+RZKkWZvPaaKNwK42vQu4tqd+f3UdBM5Jcj5wFbC/qo5X1WvAfmBDW3Z2VR1sj8y8v2dbkqRFMNAzkIEC/jJJAf+5qnYAI1X1Ulv+MjDSplcCR3rWPdpqJ6sfnab+K5JsoXu0wcjICJ1OZ8DuL7yJiYmh7n+5cJz66x2jrWsn57ydt/I4+z4azHzGadAw+J2qOpbkN4H9SX7Qu7CqqgXFKdVCaAfA6OhojY2NnepdzqjT6TDM/S8XjlN/vWN087ZH5rydwzeOLUyHliDfR4OZzzgNdJqoqo61768CD9E95/9KO8VD+/5qa34MuKBn9VWtdrL6qmnqkqRF0jcMkpyV5F1T08B64PvAHmDqiqBNwMNteg9wU7uqaB3wRjudtA9Yn2RF++B4PbCvLXszybp2FdFNPduSJC2CQU4TjQAPtas9Twf+vKq+keQx4IEkm4EXgeta+73ANcA48BPgEwBVdTzJ7cBjrd1nqup4m/4UcB9wJvD19iVJWiR9w6CqXgA+OE39R8CV09QLuGWGbe0Edk5Tfxy4eID+SpJOAf8CWZJkGEiSDANJEoaBJAnDQJKEYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkScwiDJKcluR7Sb7W5i9M8miS8SRfTXJGq7+jzY+35at7tnFbqz+X5Kqe+oZWG0+ybQFfnyRpALM5Mvg08GzP/OeAu6rqfcBrwOZW3wy81up3tXYkuQi4HvgAsAH4UguY04AvAlcDFwE3tLaSpEUyUBgkWQV8FPivbT7Ah4EHW5NdwLVtemObpy2/srXfCOyuqp9V1Q+BceDy9jVeVS9U1c+B3a2tJGmRnD5gu/8I/EvgXW3+PcDrVTXZ5o8CK9v0SuAIQFVNJnmjtV8JHOzZZu86R06oXzFdJ5JsAbYAjIyM0Ol0Buz+wpuYmBjq/pcLx6m/3jHaunby5I1P4q08zr6PBjOfceobBkn+MfBqVT2RZGxOe1kgVbUD2AEwOjpaY2PD606n02GY+18uHKf+esfo5m2PzHk7h28cW5gOLUG+jwYzn3Ea5MjgQ8DvJ7kGeCdwNvAF4Jwkp7ejg1XAsdb+GHABcDTJ6cC7gR/11Kf0rjNTXZK0CPp+ZlBVt1XVqqpaTfcD4G9W1Y3At4CPtWabgIfb9J42T1v+zaqqVr++XW10IbAG+A7wGLCmXZ10RtvHngV5dZKkgQz6mcF0/hTYneSzwPeAe1v9XuDLScaB43T/c6eqDiV5AHgGmARuqapfACS5FdgHnAbsrKpD8+iXJGmWZhUGVdUBOm36BbpXAp3Y5qfAx2dY/w7gjmnqe4G9s+mLFs/qeZzHvm/DWQvYE0mnin+BLEkyDCRJhoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkDANJEoaBJAnDQJKEYSBJwjCQJGEYSJKY38NttIzM55kEkt76+h4ZJHlnku8k+d9JDiX5t61+YZJHk4wn+Wp7ZCXtsZZfbfVHk6zu2dZtrf5ckqt66htabTzJtlPwOiVJJzHIaaKfAR+uqg8ClwAbkqwDPgfcVVXvA14DNrf2m4HXWv2u1o4kF9F9BOYHgA3Al5KcluQ04IvA1cBFwA2trSRpkfQNg+qaaLO/3r4K+DDwYKvvAq5t0xvbPG35lUnS6rur6mdV9UNgnO5jMy8Hxqvqhar6ObC7tZUkLZKBPjNov70/AbyP7m/xfwW8XlWTrclRYGWbXgkcAaiqySRvAO9p9YM9m+1d58gJ9Stm6McWYAvAyMgInU5nkO6fEhMTE0Pd/2xtXTvZv9EpsNzGaRh6x2g+/05v5XH2fTSY+YzTQGFQVb8ALklyDvAQ8Ftz2ts8VdUOYAfA6OhojY2NDaMbQPcHb5j7n62bh/QB8n0bzlpW4zQMve+l+fw7Hb5xbGE6tAQtt5+3YZnPOM3q0tKqeh34FvDbwDlJpsJkFXCsTR8DLgBoy98N/Ki3fsI6M9UlSYtkkKuJ3tuOCEhyJvAR4Fm6ofCx1mwT8HCb3tPmacu/WVXV6te3q40uBNYA3wEeA9a0q5POoPsh854FeG2SpAENcprofGBX+9zg14AHquprSZ4Bdif5LPA94N7W/l7gy0nGgeN0/3Onqg4leQB4BpgEbmmnn0hyK7APOA3YWVWHFuwVSpL66hsGVfUU8A+nqb9A90qgE+s/BT4+w7buAO6Ypr4X2DtAf7XMPH3sjTmfBz9850cXuDeSZuLtKCRJ3o5CequY7y1HPBJ7e/PIQJJkGEiSDANJEoaBJAnDQJKEYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRLem0hacLO9R9DWtZNDexKdNMUjA0mSYSBJGuyxlxck+VaSZ5IcSvLpVj83yf4kz7fvK1o9Se5OMp7kqSSX9mxrU2v/fJJNPfXLkjzd1rk7SU7Fi5UkTW+QI4NJYGtVXQSsA25JchGwDThQVWuAA20e4Gq6zzdeA2wB7oFueADbgSvoPiFt+1SAtDaf7Flvw/xfmiRpUH3DoKpeqqrvtun/AzwLrAQ2Artas13AtW16I3B/dR0EzklyPnAVsL+qjlfVa8B+YENbdnZVHayqAu7v2ZYkaRHM6mqiJKvpPg/5UWCkql5qi14GRtr0SuBIz2pHW+1k9aPT1Kfb/xa6RxuMjIzQ6XRm0/0FNTExMdT9z9bWtZND2e/ImXPf93Ia316zfb3zGaOFtJTHe7n9vA3LfMZp4DBI8veA/wH8SVW92Xtav6oqSc2pB7NQVTuAHQCjo6M1NjZ2qnc5o06nwzD3P1vDunRx69pJPv/03K5gPnzj2MJ2ZpHMdqznM0YLaSmP93L7eRuW+YzTQFcTJfl1ukHwZ1X1F638SjvFQ/v+aqsfAy7oWX1Vq52svmqauiRpkQxyNVGAe4Fnq+o/9CzaA0xdEbQJeLinflO7qmgd8EY7nbQPWJ9kRfvgeD2wry17M8m6tq+berYlSVoEgxybfgj4Q+DpJE+22r8C7gQeSLIZeBG4ri3bC1wDjAM/AT4BUFXHk9wOPNbafaaqjrfpTwH3AWcCX29fkqRF0jcMqup/ATNd93/lNO0LuGWGbe0Edk5Tfxy4uF9fJEmnhn+BLEnyRnXLxWxvfiZJs2EYaMmabwAevvOjC9QT6a3P00SSJMNAkmQYSJIwDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQGe+zlziSvJvl+T+3cJPuTPN++r2j1JLk7yXiSp5Jc2rPOptb++SSbeuqXJXm6rXN3e/SlJGkRDXJkcB+w4YTaNuBAVa0BDrR5gKuBNe1rC3APdMMD2A5cAVwObJ8KkNbmkz3rnbgvSdIpNshjL7+dZPUJ5Y3AWJveBXSAP231+9ujLw8mOSfJ+a3t/qlnHifZD2xI0gHOrqqDrX4/cC0+A1kLYD7PQ/BZCHq7mevDbUaq6qU2/TIw0qZXAkd62h1ttZPVj05Tl4bKJ8vp7WbeTzqrqkpSC9GZfpJsoXv6iZGRETqdzmLsdloTExOLuv+taycXbV8LaeTM5dv3xbJUxmiYP0/9LPbP23I1n3Gaaxi8kuT8qnqpnQZ6tdWPARf0tFvVasf45WmlqXqn1VdN035aVbUD2AEwOjpaY2NjMzU95TqdDou5/5uX6W+qW9dO8vmnfbrqySyVMTp849iwuzCjxf55W67mM05zvbR0DzB1RdAm4OGe+k3tqqJ1wBvtdNI+YH2SFe2D4/XAvrbszSTr2lVEN/VsS5K0SPr+OpLkK3R/qz8vyVG6VwXdCTyQZDPwInBda74XuAYYB34CfAKgqo4nuR14rLX7zNSHycCn6F6xdCbdD4798FiSFtkgVxPdMMOiK6dpW8AtM2xnJ7BzmvrjwMX9+iFJOnX8C2RJkmEgSTIMJEkYBpIkDANJEoaBJIkFuB2FBuf9biQtVR4ZSJIMA0mSp4kkNT7/4e3NIwNJkmEgSTIMJEkYBpIkDANJEoaBJAkvLZW0ALwsdflbMkcGSTYkeS7JeJJtw+6PJL2dLIkjgySnAV8EPgIcBR5Lsqeqnhluz37V1G9AW9dOcrP3GpLmbZCjipl+3jyqWDhLIgyAy4HxqnoBIMluYCOw5MJA0tIx35s/Gia/lO4z7IfcieRjwIaq+mdt/g+BK6rq1hPabQG2tNn3A88takf/rvOAvx3i/pcLx6k/x6g/x2gw/cbpH1TVe6dbsFSODAZSVTuAHcPuB0CSx6tqdNj9WOocp/4co/4co8HMZ5yWygfIx4ALeuZXtZokaREslTB4DFiT5MIkZwDXA3uG3CdJettYEqeJqmoyya3APuA0YGdVHRpyt/pZEqerlgHHqT/HqD/HaDBzHqcl8QGyJGm4lsppIknSEBkGkiTDYLaSfDzJoST/N8noCctua7fTeC7JVcPq41KS5N8kOZbkyfZ1zbD7tJR4G5b+khxO8nR7/zw+7P4sFUl2Jnk1yfd7aucm2Z/k+fZ9xaDbMwxm7/vAPwG+3VtMchHdq6A+AGwAvtRusyG4q6ouaV97h92ZpaLnNixXAxcBN7T3kX7V77b3j39r8Ev30f2/ptc24EBVrQEOtPmBGAazVFXPVtV0f/m8EdhdVT+rqh8C43RvsyHN5P/fhqWqfg5M3YZF6quqvg0cP6G8EdjVpncB1w66PcNg4awEjvTMH201wa1JnmqHtQMftr4N+J4ZTAF/meSJdksazWykql5q0y8DI4OuuCT+zmCpSfI/gb8/zaJ/XVUPL3Z/lrqTjRdwD3A73R/o24HPA/908Xqnt4DfqapjSX4T2J/kB+23Yp1EVVWSgf92wDCYRlX93hxWe9veUmPQ8UryX4CvneLuLCdv2/fMbFTVsfb91SQP0T29ZhhM75Uk51fVS0nOB14ddEVPEy2cPcD1Sd6R5EJgDfCdIfdp6Nobcsof0P0AXl3ehqWPJGcledfUNLAe30MnswfY1KY3AQOfyfDIYJaS/AHwn4D3Ao8kebKqrqqqQ0keoPsMhknglqr6xTD7ukT8uySX0D1NdBj4o6H2ZglZprdhWWwjwENJoPv/1Z9X1TeG26WlIclXgDHgvCRHge3AncADSTYDLwLXDbw9b0chSfI0kSTJMJAkGQaSJAwDSRKGgSQJw0CShGEgSQL+H2DBMsBWtT9AAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "protein_data['score'].hist(bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "designed                                56326\n",
       "random_doubles                          25040\n",
       "rnn_standard_walked                     20838\n",
       "cnn_designed_plus_rand_train_walked     20759\n",
       "rnn_designed_plus_rand_train_walked     20731\n",
       "lr_standard_walked                      20456\n",
       "cnn_rand_doubles_plus_single_walked     20454\n",
       "cnn_standard_walked                     20395\n",
       "rnn_rand_doubles_plus_singles_walked    20154\n",
       "lr_rand_doubles_plus_single_walked      19999\n",
       "lr_designed_plus_rand_train_walked      19680\n",
       "rand                                     9857\n",
       "lr_rand_doubles_plus_single_seed         2071\n",
       "rnn_designed_plus_rand_train_seed        2065\n",
       "rnn_rand_doubles_plus_singles_seed       2045\n",
       "lr_designed_plus_rand_train_seed         2030\n",
       "cnn_rand_doubles_plus_single_seed        2022\n",
       "lr_standard_seed                         1989\n",
       "cnn_standard_seed                        1924\n",
       "rnn_standard_seed                        1916\n",
       "cnn_designed_plus_rand_train_seed        1898\n",
       "single                                   1112\n",
       "singles                                  1085\n",
       "previous_chip_nonviable                   998\n",
       "previous_chip_viable                      994\n",
       "stop                                       57\n",
       "wild_type                                   1\n",
       "Name: partition, dtype: int64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aav_data.partition.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4     9670\n",
       "3     9340\n",
       "5     9153\n",
       "6     8514\n",
       "2     5800\n",
       "7     1289\n",
       "8     1215\n",
       "9     1142\n",
       "10    1073\n",
       "11    1011\n",
       "12     965\n",
       "13     882\n",
       "14     816\n",
       "15     736\n",
       "16     653\n",
       "17     578\n",
       "18     526\n",
       "19     450\n",
       "20     393\n",
       "21     356\n",
       "22     289\n",
       "23     259\n",
       "24     211\n",
       "25     184\n",
       "26     148\n",
       "27     118\n",
       "28     103\n",
       "29      93\n",
       "30      85\n",
       "31      63\n",
       "32      50\n",
       "33      47\n",
       "34      41\n",
       "35      27\n",
       "36      17\n",
       "38      13\n",
       "37      13\n",
       "39       3\n",
       "Name: num_edits, dtype: int64"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(aav_data.query(\"partition == 'designed'\")['num_edits'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "designed                                56326\n",
       "random_doubles                          25040\n",
       "rnn_standard_walked                     20838\n",
       "cnn_designed_plus_rand_train_walked     20759\n",
       "rnn_designed_plus_rand_train_walked     20731\n",
       "lr_standard_walked                      20456\n",
       "cnn_rand_doubles_plus_single_walked     20454\n",
       "cnn_standard_walked                     20395\n",
       "rnn_rand_doubles_plus_singles_walked    20154\n",
       "lr_rand_doubles_plus_single_walked      19999\n",
       "lr_designed_plus_rand_train_walked      19680\n",
       "rand                                     9857\n",
       "lr_rand_doubles_plus_single_seed         2071\n",
       "rnn_designed_plus_rand_train_seed        2065\n",
       "rnn_rand_doubles_plus_singles_seed       2045\n",
       "lr_designed_plus_rand_train_seed         2030\n",
       "cnn_rand_doubles_plus_single_seed        2022\n",
       "lr_standard_seed                         1989\n",
       "cnn_standard_seed                        1924\n",
       "rnn_standard_seed                        1916\n",
       "cnn_designed_plus_rand_train_seed        1898\n",
       "single                                   1112\n",
       "singles                                  1085\n",
       "previous_chip_nonviable                   998\n",
       "previous_chip_viable                      994\n",
       "stop                                       57\n",
       "wild_type                                   1\n",
       "Name: category, dtype: int64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_data.category.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "design_train = [\n",
    "    'previous_chip_viable',\n",
    "    'previous_chip_nonviable', \n",
    "    'stop',\n",
    "    'single',\n",
    "    'singles',\n",
    "    'wild_type',\n",
    "    'random_doubles'\n",
    "]\n",
    "\n",
    "design_test = [\n",
    "    'designed',\n",
    "    'cnn_designed_plus_rand_train_walked',\n",
    "    'lr_rand_doubles_plus_single_walked',\n",
    "    'cnn_rand_doubles_plus_single_seed',\n",
    "    'cnn_standard_walked',\n",
    "    'rnn_rand_doubles_plus_singles_seed',\n",
    "    'lr_standard_walked',\n",
    "    'lr_rand_doubles_plus_single_seed',\n",
    "    'lr_designed_plus_rand_train_seed',\n",
    "    'cnn_rand_doubles_plus_single_walked',\n",
    "    'cnn_designed_plus_rand_train_seed',\n",
    "    'lr_standard_seed',\n",
    "    'rnn_designed_plus_rand_train_walked',\n",
    "    'lr_designed_plus_rand_train_walked',\n",
    "    'rnn_standard_seed',\n",
    "    'rnn_standard_walked',\n",
    "    'rnn_designed_plus_rand_train_seed',\n",
    "    'cnn_standard_seed',\n",
    "    'rnn_rand_doubles_plus_singles_walked',\n",
    "]\n",
    "\n",
    "def design_split(category):\n",
    "    if category in design_train:\n",
    "        return \"train\"\n",
    "    elif category in design_test:\n",
    "        return \"test\"\n",
    "    else:\n",
    "        return \"unknown\"\n",
    "    \n",
    "def design_split_reversed(category):\n",
    "    if category in design_train:\n",
    "        return \"test\"\n",
    "    elif category in design_test:\n",
    "        return \"train\"\n",
    "    else:\n",
    "        return \"unknown\"\n",
    "\n",
    "protein_data['design_split'] = protein_data.category.apply(design_split)\n",
    "protein_data['design_split_reversed'] = protein_data.category.apply(design_split_reversed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The test set will be 63.866471761155424% of the data.\n"
     ]
    }
   ],
   "source": [
    "natural_allowed = [\n",
    "    'previous_chip_viable',\n",
    "    'previous_chip_nonviable', \n",
    "    'stop',\n",
    "    'single',\n",
    "    'singles',\n",
    "    'wild_type',\n",
    "    'random_doubles'\n",
    "]\n",
    "\n",
    "# randomly pick a few, see distribution of mutations\n",
    "def natural_split_1(category):\n",
    "    if category in natural_allowed:\n",
    "        return random.choices(['train', 'test'], weights=[0.8, 0.2], k=1)[0]\n",
    "    else:\n",
    "        return \"unknown\"\n",
    "\n",
    "# from train, those w/ many mutations\n",
    "\n",
    "num_mutations_threshold = 7\n",
    "\n",
    "test_count = len(protein_data[protein_data.number_of_mutations > num_mutations_threshold])\n",
    "total = len(protein_data)\n",
    "\n",
    "test_percentage = (test_count*100)/total\n",
    "\n",
    "print(f'The test set will be {test_percentage}% of the data.')\n",
    "\n",
    "def natural_split_2(row):\n",
    "    if row.category in natural_allowed:\n",
    "        if row.number_of_mutations > num_mutations_threshold:\n",
    "            return \"test\"\n",
    "        else:\n",
    "            return \"train\"\n",
    "    else:\n",
    "        return \"unknown\"\n",
    "    \n",
    "\n",
    "protein_data['natural_split_1'] = protein_data.category.apply(natural_split_1)\n",
    "protein_data['natural_split_2'] = protein_data.apply(natural_split_2, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Random sample n sequences from train k times for single shot learning approaches\n",
    "\n",
    "n = 96\n",
    "k = 100\n",
    "\n",
    "protein_data['natural_split_1_random_train_set'] = 'unassigned'\n",
    "protein_data['natural_split_2_random_train_set'] = 'unassigned'\n",
    "\n",
    "natural_split_1_train = protein_data.query('natural_split_1 == \"train\"')\n",
    "natural_split_1_train_index = list(natural_split_1_train.index)\n",
    "\n",
    "natural_split_1_kn_indexes = numpy.array(random.sample(natural_split_1_train_index, k=k*n)).reshape(k,n)\n",
    "\n",
    "for i, kn_index in enumerate(natural_split_1_kn_indexes):\n",
    "    protein_data.loc[kn_index, 'natural_split_1_random_train_set'] = f\"S{i+1}\"\n",
    "    \n",
    "natural_split_2_train = protein_data.query('natural_split_2 == \"train\"')\n",
    "natural_split_2_train_index = list(natural_split_2_train.index)\n",
    "\n",
    "natural_split_2_kn_indexes = numpy.array(random.sample(natural_split_2_train_index, k=k*n)).reshape(k,n)\n",
    "\n",
    "for i, kn_index in enumerate(natural_split_2_kn_indexes):\n",
    "    protein_data.loc[kn_index, 'natural_split_2_random_train_set'] = f\"S{i+1}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Set design_task has 29287 train and 257752 test sequences. 57% of test is positive.\n",
      "    \n",
      "\n",
      "    Set design_task_reversed has 257752 train and 29287 test sequences. 23% of test is positive.\n",
      "    \n",
      "\n",
      "    Set natural_task_1 has 23466 train and 5821 test sequences. 24% of test is positive.\n",
      "    \n",
      "\n",
      "    Set natural_task_2 has 28845 train and 442 test sequences. 22% of test is positive.\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "# Check how much are in test/train for each set\n",
    "\n",
    "def get_stats(split_name):\n",
    "    return f\"\"\"\n",
    "    Set {split_name} has {\n",
    "      len(protein_data.query(f\"{split_name}=='train'\"))\n",
    "    } train and {\n",
    "      len(protein_data.query(f\"{split_name}=='test'\"))\n",
    "    } test sequences. {\n",
    "      round(len(protein_data.query(f\"{split_name}=='test' & binary_score==True\")) /\n",
    "      len(protein_data.query(f\"{split_name}=='test'\")) * 100)\n",
    "    }% of test is positive.\n",
    "    \"\"\"\n",
    "\n",
    "print(get_stats(\"design_split\"))\n",
    "print(get_stats(\"design_split_reversed\"))\n",
    "print(get_stats(\"natural_split_1\"))\n",
    "print(get_stats(\"natural_split_2\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add reference region for completeness\n",
    "protein_data['reference_region'] = reference_region\n",
    "\n",
    "protein_data[\n",
    "    ['mutation_mask', 'mutated_region', 'category', 'number_of_mutations',\n",
    "    'levenshtein_distance', 'score', 'full_aa_sequence', 'binary_score', 'reference_region',\n",
    "    'design_split', 'design_split_reversed', 'natural_split_1', 'natural_split_2',\n",
    "    'natural_split_1_random_train_set', 'natural_split_2_random_train_set']\n",
    "].to_csv(split_path / \"full_data.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "ml_categories = ['train','test']\n",
    "\n",
    "# Design regression\n",
    "design_split_regression = protein_data[protein_data.design_split.isin(ml_categories)]\n",
    "design_split_regression = design_split_regression.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'score': 'target',\n",
    "    'design_split': 'set',\n",
    "}, axis=1)\n",
    "\n",
    "design_split_regression[\n",
    "    ['sequence', 'target', 'set']\n",
    "].to_csv(split_path / 'splits' / \"design_split_regression.csv\", index=False)\n",
    "\n",
    "del design_split_regression\n",
    "\n",
    "# Design reversed regression\n",
    "design_split_reversed_regression = protein_data[protein_data.design_split_reversed.isin(ml_categories)]\n",
    "design_split_reversed_regression = design_split_reversed_regression.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'score': 'target',\n",
    "    'design_split_reversed': 'set',\n",
    "}, axis=1)\n",
    "\n",
    "design_split_reversed_regression[\n",
    "    ['sequence', 'target', 'set']\n",
    "].to_csv(split_path / 'splits' / \"design_split_reversed_regression.csv\", index=False)\n",
    "\n",
    "del design_split_reversed_regression\n",
    "\n",
    "# Natural 1 regression\n",
    "natural_split_1_regression = protein_data[protein_data.natural_split_1.isin(ml_categories)]\n",
    "natural_split_1_regression = natural_split_1_regression.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'score': 'target',\n",
    "    'natural_split_1': 'set',\n",
    "    'natural_split_1_random_train_set': 'random_sample'\n",
    "}, axis=1)\n",
    "\n",
    "natural_split_1_regression[\n",
    "    ['sequence', 'target', 'set', 'random_sample']\n",
    "].to_csv(split_path / 'splits' / \"natural_split_1_regression.csv\", index=False)\n",
    "\n",
    "del natural_split_1_regression\n",
    "\n",
    "# Natural 2 regression\n",
    "natural_split_2_regression = protein_data[protein_data.natural_split_2.isin(ml_categories)]\n",
    "natural_split_2_regression = natural_split_2_regression.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'score': 'target',\n",
    "    'natural_split_2': 'set',\n",
    "    'natural_split_2_random_train_set': 'random_sample'\n",
    "}, axis=1)\n",
    "\n",
    "natural_split_2_regression[\n",
    "    ['sequence', 'target', 'set', 'random_sample']\n",
    "].to_csv(split_path / 'splits' / \"natural_split_2_regression.csv\", index=False)\n",
    "\n",
    "del natural_split_2_regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}