{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notes\n",
    "\n",
    "- Data taken from here: https://github.com/churchlab/Deep_diversification_AAV/tree/main/Data\n",
    "- Just sequence (AA) data exists as well here: https://www.dropbox.com/sh/lmd8wmgibz24r2h/AADmgvc_0Q5mylwgGf97cTC-a/ML_and_Analysis/model_and_dataset_analysis/allseqs_20191230.csv.zip?dl=0\n",
    "- [This notebook](https://github.com/churchlab/Deep_diversification_AAV/blob/main/Parsing_pipeline/Step3_compute_selection_scores.ipynb) suggest that the \"target\" column is `GAS1_virus_S` in the CSV file\n",
    "- [This repository](https://github.com/google-research/google-research/tree/master/aav/model_training) suggest that the wild type AA sequence is 28 residues and is `DEEEIRTTNPVATEQYGSVSTNLQRGNR`\n",
    "- Cutoff via gaussian mixture copied from: https://colab.research.google.com/github/google-research/google-research/blob/master/aav/model_and_dataset_analysis/data_prep.ipynb\n",
    "- The UniProt sequence is here: https://www.uniprot.org/uniprot/P03135\n",
    "- In UniProt notation (sequence starts at 1), the region on P03135 where the sequence mutates is `[561, 588]`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (1.78)\n",
      "Requirement already satisfied: numpy in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from biopython) (1.20.2)\n",
      "Requirement already satisfied: pandas in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (1.2.4)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from pandas) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.3 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from pandas) (2021.1)\n",
      "Requirement already satisfied: numpy>=1.16.5 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from pandas) (1.20.2)\n",
      "Requirement already satisfied: six>=1.5 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from python-dateutil>=2.7.3->pandas) (1.16.0)\n",
      "Requirement already satisfied: matplotlib in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (3.4.2)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (2.4.7)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (0.10.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (2.8.1)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (8.2.0)\n",
      "Requirement already satisfied: numpy>=1.16 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (1.20.2)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from matplotlib) (1.3.1)\n",
      "Requirement already satisfied: six in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from cycler>=0.10->matplotlib) (1.16.0)\n",
      "Requirement already satisfied: scikit-learn in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (0.24.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from scikit-learn) (2.1.0)\n",
      "Requirement already satisfied: scipy>=0.19.1 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from scikit-learn) (1.6.3)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from scikit-learn) (1.20.2)\n",
      "Requirement already satisfied: joblib>=0.11 in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from scikit-learn) (1.0.1)\n",
      "Requirement already satisfied: python-Levenshtein in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (0.12.2)\n",
      "Requirement already satisfied: setuptools in /Users/chdallago/miniconda3/envs/bio-benchmarks/lib/python3.8/site-packages (from python-Levenshtein) (49.6.0.post20210108)\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython\n",
    "!pip install pandas\n",
    "!pip install matplotlib\n",
    "!pip install scikit-learn\n",
    "!pip install python-Levenshtein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import math\n",
    "import random\n",
    "import Levenshtein\n",
    "import numpy\n",
    "\n",
    "from pathlib import Path\n",
    "from pandas import read_csv, DataFrame\n",
    "\n",
    "from helpers import read_fasta, TwoGaussianMixtureModelLabeler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pandas.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = Path('') / '..' / 'data' / 'aav'\n",
    "aav_data_path = data_path / 'library_w_selection_scores.csv'\n",
    "\n",
    "task_path = Path('') / '..' / 'tasks' / 'aav'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "aav_data = read_csv(aav_data_path, index_col=0)\n",
    "\n",
    "P03135 = read_fasta(data_path / 'P03135.fasta')[0]\n",
    "region = (561,588)\n",
    "reference_region = \"DEEEIRTTNPVATEQYGSVSTNLQRGNR\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['nt_seq', 'EK266_GAS1_p1_rep1a_plasmid', 'EK266_GAS1_p1_rep1b_plasmid',\n",
       "       'EK269_GAS1_p1_rep1a_plasmid_x', 'EK269_GAS1_p1_rep1b_plasmid_x',\n",
       "       'EK269_GAS1_p1_rep1c_plasmid_x', 'EK269_GAS1_p1_rep1d_plasmid_x',\n",
       "       'EK269_GAS1_v3_rep1a_virus_x', 'EK269_GAS1_v3_rep1b_virus_x',\n",
       "       'EK269_GAS1_v3_rep1c_virus_x', 'EK269_GAS1_v3_rep1d_virus_x',\n",
       "       'EK269_GAS1_v4_rep2a_virus_x', 'EK269_GAS1_v4_rep2b_virus_x',\n",
       "       'EK269_GAS1_v4_rep2c_virus_x', 'EK269_GAS1_v4_rep2d_virus_x',\n",
       "       'EK269_GAS1_v5_rep3a_virus_x', 'EK269_GAS1_v5_rep3b_virus_x',\n",
       "       'EK269_GAS1_v5_rep3c_virus', 'EK269_GAS1_v5_rep3d_virus', 'aa_x',\n",
       "       'category_x', 'chip_x', 'control_x', 'is_wt_aa_x', 'is_wt_nt_x',\n",
       "       'mask_x', 'mut_x', 'rep_i', 'rep_original_x', 'rep_total_x',\n",
       "       'EK269_GAS1_p1_rep1a_plasmid_y', 'EK269_GAS1_p1_rep1b_plasmid_y',\n",
       "       'EK269_GAS1_p1_rep1c_plasmid_y', 'EK269_GAS1_p1_rep1d_plasmid_y',\n",
       "       'EK269_GAS1_v3_rep1a_virus_y', 'EK269_GAS1_v3_rep1b_virus_y',\n",
       "       'EK269_GAS1_v3_rep1c_virus_y', 'EK269_GAS1_v3_rep1d_virus_y',\n",
       "       'EK269_GAS1_v4_rep2a_virus_y', 'EK269_GAS1_v4_rep2b_virus_y',\n",
       "       'EK269_GAS1_v4_rep2c_virus_y', 'EK269_GAS1_v4_rep2d_virus_y',\n",
       "       'EK269_GAS1_v5_rep3a_virus_y', 'EK269_GAS1_v5_rep3b_virus_y', 'aa_y',\n",
       "       'category_y', 'chip_y', 'control_y', 'is_wt_aa_y', 'is_wt_nt_y',\n",
       "       'mask_y', 'mut_y', 'rep_original_y', 'rep_total_y', 'GAS1_plasmid_N',\n",
       "       'GAS1_virus_N', 'GAS1_plasmid_F', 'GAS1_virus_F', 'GAS1_virus_S'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aav_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['rnn_designed_plus_rand_train_walked',\n",
       "       'rnn_designed_plus_rand_train_seed',\n",
       "       'rnn_rand_doubles_plus_singles_walked',\n",
       "       'rnn_rand_doubles_plus_singles_seed', 'rnn_standard_walked',\n",
       "       'rnn_standard_seed', 'cnn_designed_plus_rand_train_walked',\n",
       "       'cnn_designed_plus_rand_train_seed',\n",
       "       'cnn_rand_doubles_plus_single_walked',\n",
       "       'cnn_rand_doubles_plus_single_seed', 'cnn_standard_walked',\n",
       "       'cnn_standard_seed', 'lr_designed_plus_rand_train_walked',\n",
       "       'lr_designed_plus_rand_train_seed',\n",
       "       'lr_rand_doubles_plus_single_walked',\n",
       "       'lr_rand_doubles_plus_single_seed', 'lr_standard_walked',\n",
       "       'lr_standard_seed', 'previous_chip_viable',\n",
       "       'previous_chip_nonviable', 'stop', 'singles', 'wild_type',\n",
       "       'random_doubles'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aav_data['category_x'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the wild type sequence from the data\n",
    "\n",
    "wt = aav_data[aav_data['is_wt_nt_x'] == 1].to_dict(orient='records')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure that all assumptions on the wild type are correct\n",
    "\n",
    "assert(wt['aa_x'] == reference_region)\n",
    "\n",
    "# Important: region[0] -1 because UniProt notation! (0 in Python == 1 in UniProt)\n",
    "assert(wt['aa_x'] == str(P03135[region[0]-1 : region[1]].seq))\n",
    "\n",
    "# Make sure chopping the sequence works as expected:\n",
    "\n",
    "assert(\n",
    "    str(P03135.seq) ==\n",
    "    str(P03135[:region[0]-1].seq) + str(P03135[region[0]-1 : region[1]].seq) + str(P03135[region[1]:].seq)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_data = aav_data[['aa_x', 'mask_x', 'GAS1_virus_S', 'category_x', 'mut_x']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove data where GAS1_virus_S is +/- inf\n",
    "#protein_data = protein_data[protein_data['GAS1_virus_S'] != math.inf]\n",
    "#protein_data = protein_data[protein_data['GAS1_virus_S'] != -math.inf]\n",
    "\n",
    "protein_data.replace([math.inf, -math.inf], math.nan, inplace=True)\n",
    "protein_data.dropna(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We need to aggregate synonymous variants if operating on AA sequences\n",
    "# Take the score and devide by number of synonymous WT AA seuqneces!\n",
    "\n",
    "def aggregate_rows(dataframe_slice):\n",
    "    \n",
    "    first_element = dataframe_slice.iloc[0]\n",
    "    \n",
    "    return {\n",
    "        'mutation_mask': first_element.mask_x,\n",
    "        'mutated_region': first_element.aa_x,\n",
    "        'category': first_element.category_x,\n",
    "        'number_of_mutations': first_element.mut_x,\n",
    "        'levenshtein_distance': Levenshtein.distance(first_element.aa_x, reference_region),\n",
    "        'score': dataframe_slice.GAS1_virus_S.mean()\n",
    "    }\n",
    "\n",
    "grouped_protein_data = protein_data.groupby(['aa_x']).apply(aggregate_rows).values\n",
    "\n",
    "protein_data = DataFrame.from_records(grouped_protein_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def region_cut(new_sequence):\n",
    "    filtered_sequence = new_sequence.upper().replace('*','')\n",
    "    return str(P03135[:region[0]-1].seq) + filtered_sequence + str(P03135[region[1]:].seq)\n",
    "\n",
    "protein_data['full_aa_sequence'] = protein_data['mutated_region'].apply(region_cut)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the the manuscript:\n",
    " \n",
    "> To avoid overfitting to experimental noise, rather than predicting the quantitative production efficiency we used binary classification models to predict whether each sequence variant is viable (Supplementary Fig. 1), as defined by a threshold fit to best separate positive and negative controls (WT replicas, and variants containing stop codons, respectively).\n",
    "\n",
    "Apparently, by plotting `GAS1_virus_S` one should see a binomial. The cutoff is set somewhere beteen -2.5 and 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYMAAAD4CAYAAAAO9oqkAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUDUlEQVR4nO3df6zd9X3f8eerdiAIkpiE9JbaaCaq1c0JagpX4K6pdBMyMBDVdEoiEComZbG6gJRKTI2zaKULiUQ0tdmQkrTesAJRVoelZVjEzPEIV9H+MMEk/AgQxg1xhi0KakygTlYid+/9cT6Ojm7O8T331zmH6+dDOrrf7+f7+X7P+3zP1/d1vz+dqkKSdHL7pVEXIEkaPcNAkmQYSJIMA0kShoEkCVg96gIW6qyzzqr169ePuoy+fvKTn3D66aePuox5s+7hsu7hOtnrfvjhh/+uqt7aa9prNgzWr1/PgQMHRl1GX9PT00xNTY26jHmz7uGy7uE62etO8sN+0zxMJEkyDCRJhoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkXsN3IEsaH+u3f23B8x689YolrEQL5Z6BJMkwkCQZBpIkDANJEoaBJAnDQJKEYSBJwjCQJGEYSJIYMAySHEzyeJJHkhxobW9Osi/JM+3nma09SW5LMpPksSTndy1na+v/TJKtXe0XtOXPtHmz1B9UktTffPYM3l1V76yqyTa+Hbi/qjYA97dxgMuADe21DfgCdMIDuBm4CLgQuPl4gLQ+H+6ab/OCP5Ekad4Wc5hoC3BHG74DuLKr/c7q2A+sSXI2cCmwr6qOVNVLwD5gc5v2xqraX1UF3Nm1LEnSEAz6oLoCvp6kgL+sqh3ARFU936b/LTDRhtcCz3XNe6i1naj9UI/2X5BkG529DSYmJpienh6w/OE7evToWNfXj3UP10qp+6bzji14WcP8/CtlfS+HQcPgXVV1OMkvA/uSfK97YlVVC4pl1UJoB8Dk5GRNTU0t91su2PT0NONcXz/WPVwrpe7rFvPU0mum5uyzVFbK+l4OAx0mqqrD7eeLwN10jvm/0A7x0H6+2LofBs7pmn1daztR+7oe7ZKkIZkzDJKcnuQNx4eBS4DvAruB41cEbQXuacO7gWvbVUWbgJfb4aS9wCVJzmwnji8B9rZpryTZ1K4iurZrWZKkIRjkMNEEcHe72nM18F+r6n8keQi4K8n1wA+BD7b+e4DLgRngp8CHAKrqSJJbgIdav09W1ZE2/BHgi8BpwH3tJUkakjnDoKqeBX6jR/uPgIt7tBdwQ59l7QR29mg/ALxjgHolScvAO5AlSYaBJMkwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkDANJEoaBJAnDQJKEYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiTmEQZJViX5TpJ72/i5SR5MMpPkK0lOae2ntvGZNn191zI+3tqfTnJpV/vm1jaTZPsSfj5J0gDms2fwUeCprvHPAJ+tql8DXgKub+3XAy+19s+2fiTZCFwFvB3YDHy+Bcwq4HPAZcBG4OrWV5I0JAOFQZJ1wBXAf2njAd4DfLV1uQO4sg1vaeO06Re3/luAXVX1alX9AJgBLmyvmap6tqp+BuxqfSVJQ7J6wH7/Efhj4A1t/C3Aj6vqWBs/BKxtw2uB5wCq6liSl1v/tcD+rmV2z/PcrPaLehWRZBuwDWBiYoLp6ekByx++o0ePjnV9/Vj3cK2Uum8671j/znMY5udfKet7OcwZBkneB7xYVQ8nmVrWauZQVTuAHQCTk5M1NTXSck5oenqaca6vH+serpVS93Xbv7bgZR28ZmrOPktlpazv5TDInsFvA7+b5HLg9cAbgf8ErEmyuu0drAMOt/6HgXOAQ0lWA28CftTVflz3PP3aJUlDMOc5g6r6eFWtq6r1dE4Af6OqrgEeAN7fum0F7mnDu9s4bfo3qqpa+1XtaqNzgQ3At4CHgA3t6qRT2nvsXpJPJ0kayKDnDHr5GLAryaeA7wC3t/bbgS8lmQGO0PnlTlU9keQu4EngGHBDVf0jQJIbgb3AKmBnVT2xiLokSfM0rzCoqmlgug0/S+dKoNl9/gH4QJ/5Pw18ukf7HmDPfGqRJC0d70CWJBkGkiTDQJKEYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEkCVo+6AEknt/Xbv7bgeQ/eesUSVnJyc89AkmQYSJIMA0kShoEkCcNAkoRhIEnCMJAkYRhIkhggDJK8Psm3kjya5Ikk/761n5vkwSQzSb6S5JTWfmobn2nT13ct6+Ot/ekkl3a1b25tM0m2L8PnlCSdwCB7Bq8C76mq3wDeCWxOsgn4DPDZqvo14CXg+tb/euCl1v7Z1o8kG4GrgLcDm4HPJ1mVZBXwOeAyYCNwdesrSRqSOcOgOo620de1VwHvAb7a2u8ArmzDW9o4bfrFSdLad1XVq1X1A2AGuLC9Zqrq2ar6GbCr9ZUkDclA5wzaX/CPAC8C+4DvAz+uqmOtyyFgbRteCzwH0Ka/DLylu33WPP3aJUlDMtCD6qrqH4F3JlkD3A380+Usqp8k24BtABMTE0xPT4+ijIEcPXp0rOvrx7qHa6XUfdN5x/p3XkbzXXcrZX0vh3k9tbSqfpzkAeC3gDVJVre//tcBh1u3w8A5wKEkq4E3AT/qaj+ue55+7bPffwewA2BycrKmpqbmU/5QTU9PM8719WPdw7VS6r5uEU8eXYyD10zN2afbSlnfy2GQq4ne2vYISHIa8C+Ap4AHgPe3bluBe9rw7jZOm/6NqqrWflW72uhcYAPwLeAhYEO7OukUOieZdy/BZ5MkDWiQPYOzgTvaVT+/BNxVVfcmeRLYleRTwHeA21v/24EvJZkBjtD55U5VPZHkLuBJ4BhwQzv8RJIbgb3AKmBnVT2xZJ9QkjSnOcOgqh4DfrNH+7N0rgSa3f4PwAf6LOvTwKd7tO8B9gxQryRpGXgHsiTJMJAkGQaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkDANJEoaBJAnDQJKEYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSQJWD3qArSyrd/+tQXPe/DWK5awEkknMueeQZJzkjyQ5MkkTyT5aGt/c5J9SZ5pP89s7UlyW5KZJI8lOb9rWVtb/2eSbO1qvyDJ422e25JkOT6sJKm3QQ4THQNuqqqNwCbghiQbge3A/VW1Abi/jQNcBmxor23AF6ATHsDNwEXAhcDNxwOk9flw13ybF//RJEmDmjMMqur5qvp2G/574ClgLbAFuKN1uwO4sg1vAe6sjv3AmiRnA5cC+6rqSFW9BOwDNrdpb6yq/VVVwJ1dy5IkDUE6v38H7JysB74JvAP4P1W1prUHeKmq1iS5F7i1qv5Xm3Y/8DFgCnh9VX2qtf874P8C063/e1v77wAfq6r39Xj/bXT2NpiYmLhg165d8//EQ3L06FHOOOOMUZcxb0td9+OHX17wvOetfdPAfV3fwzW77sV8z4sxn20EVs76Xqh3v/vdD1fVZK9pA59ATnIG8NfAH1XVK92H9auqkgyeKgtUVTuAHQCTk5M1NTW13G+5YNPT04xzff0sdd3XLeYE8jWD1+H6Hq7ZdS/me16M+WwjsHLW93IYKAySvI5OEHy5qv6mNb+Q5Oyqer4d6nmxtR8GzumafV1rO0xn76C7fbq1r+vRXye5+VyJdNN5x37hF5JXI0mDG+RqogC3A09V1Z93TdoNHL8iaCtwT1f7te2qok3Ay1X1PLAXuCTJme3E8SXA3jbtlSSb2ntd27UsSdIQDLJn8NvA7wOPJ3mktf1b4FbgriTXAz8EPtim7QEuB2aAnwIfAqiqI0luAR5q/T5ZVUfa8EeALwKnAfe1lyRpSOYMg3YiuN91/xf36F/ADX2WtRPY2aP9AJ2T0pKkEfBxFJIkw0CSZBhIkjAMJEkYBpIkDANJEoaBJAnDQJKE/9OZBrCY/61M0muDewaSJMNAkmQYSJIwDCRJGAaSJAwDSRKGgSQJw0CShGEgScI7kCU187nT/KbzjnGdd6avKO4ZSJIMA0mSYSBJwnMGWsEW87TVg7desYSVSOPPPQNJkmEgSTIMJEkYBpIkDANJEoaBJAnDQJLEAGGQZGeSF5N8t6vtzUn2JXmm/TyztSfJbUlmkjyW5Pyueba2/s8k2drVfkGSx9s8tyXJUn9ISdKJDbJn8EVg86y27cD9VbUBuL+NA1wGbGivbcAXoBMewM3ARcCFwM3HA6T1+XDXfLPfS5K0zOa8A7mqvplk/azmLcBUG74DmAY+1trvrKoC9idZk+Ts1ndfVR0BSLIP2JxkGnhjVe1v7XcCVwL3LeZD6RcNejeuT6OUTk4LfRzFRFU934b/Fphow2uB57r6HWptJ2o/1KO9pyTb6OxxMDExwfT09ALLX35Hjx4dq/puOu/YQP0mThu87zhZ6rqH9d2N03Yyn/U3LtvJfNfdOK3v+RhG3Yt+NlFVVZJaimIGeK8dwA6AycnJmpqaGsbbLsj09DTjVN+gf+3fdN4x/uzx194jq5a67oPXTC3Zsk5knLaT+ewRjst2Mt/vaZzW93wMo+6FXk30Qjv8Q/v5Yms/DJzT1W9daztR+7oe7ZKkIVpoGOwGjl8RtBW4p6v92nZV0Sbg5XY4aS9wSZIz24njS4C9bdorSTa1q4iu7VqWJGlI5tzPS/JXdE4An5XkEJ2rgm4F7kpyPfBD4IOt+x7gcmAG+CnwIYCqOpLkFuCh1u+Tx08mAx+hc8XSaXROHHvyWJKGbJCria7uM+niHn0LuKHPcnYCO3u0HwDeMVcdkqTl4x3IkiTDQJJkGEiS8P9Alnry/0/WycY9A0mSYSBJMgwkSRgGkiQMA0kShoEkCcNAkoT3GUgrxmLujZDcM5AkuWcgLbX5/IU++/+c9u5ljYph8BrhIQBJy8kwkMaIoa9R8ZyBJMkwkCQZBpIkDANJEoaBJAmvJpL0Gjbfq6+8r6M/9wwkSYaBJMkwkCRhGEiSMAwkSXg10VD53BlJ48o9A0mSYSBJGqMwSLI5ydNJZpJsH3U9knQyGYswSLIK+BxwGbARuDrJxtFWJUknj3E5gXwhMFNVzwIk2QVsAZ4caVU9DHoSePZt75LGz2Iu6lhpj7JIVY26BpK8H9hcVf+qjf8+cFFV3Tir3zZgWxv9deDpoRY6P2cBfzfqIhbAuofLuofrZK/7n1TVW3tNGJc9g4FU1Q5gx6jrGESSA1U1Oeo65su6h8u6h8u6+xuLcwbAYeCcrvF1rU2SNATjEgYPARuSnJvkFOAqYPeIa5Kkk8ZYHCaqqmNJbgT2AquAnVX1xIjLWqzXxOGsHqx7uKx7uKy7j7E4gSxJGq1xOUwkSRohw0CSZBgsRpIPJHkiyf9LMjlr2sfbozWeTnJpn/nPTfJg6/eVdvJ8qNr7PtJeB5M80qffwSSPt34Hhlxmr3r+NMnhrtov79NvrB5zkuQ/JPlekseS3J1kTZ9+Y7G+51p/SU5t29BM25bXj6DM2TWdk+SBJE+2f58f7dFnKsnLXdvPn4yi1tnm+t7TcVtb348lOX/J3ryqfC3wBfwzOje/TQOTXe0bgUeBU4Fzge8Dq3rMfxdwVRv+C+Bfj/jz/BnwJ32mHQTOGvU676rnT4F/M0efVW3dvw04pX0nG0dc9yXA6jb8GeAz47q+B1l/wEeAv2jDVwFfGYNt42zg/Db8BuB/96h7Crh31LXO93sHLgfuAwJsAh5cqvd2z2ARquqpqup1F/QWYFdVvVpVPwBm6Dxy4+eSBHgP8NXWdAdw5TKWe0Ktng8CfzWqGpbBzx9zUlU/A44/5mRkqurrVXWsje6nc0/NuBpk/W2hs+1CZ1u+uG1LI1NVz1fVt9vw3wNPAWtHWdMS2gLcWR37gTVJzl6KBRsGy2Mt8FzX+CF+cWN8C/Djrl8MvfoM0+8AL1TVM32mF/D1JA+3x4KMgxvbrvLOJGf2mD7I9zBKf0Dnr7xexmF9D7L+ft6nbcsv09m2x0I7bPWbwIM9Jv9WkkeT3Jfk7cOtrK+5vvdl26bH4j6DcZbkfwK/0mPSJ6rqnmHXsxADfoarOfFewbuq6nCSXwb2JfleVX1zqWvtdqK6gS8At9D5x3MLnUNcf7Cc9QxqkPWd5BPAMeDLfRYz9PW90iQ5A/hr4I+q6pVZk79N5zk9R9v5pv8ObBhyib2M7Hs3DOZQVe9dwGyDPF7jR3R28Va3v6iW7REcc32GJKuBfwlccIJlHG4/X0xyN51DCMu6kQ667pP8Z+DeHpNG8piTAdb3dcD7gIurHQjusYyhr+8eBll/x/scatvRm+hs2yOV5HV0guDLVfU3s6d3h0NV7Uny+SRnVdVIH2I3wPe+bNu0h4mWx27gqnalxbl0/uL4VneH9kvgAeD9rWkrMKo9jfcC36uqQ70mJjk9yRuOD9M5CfrdIdbXq6bu46S/R+96xu4xJ0k2A38M/G5V/bRPn3FZ34Osv910tl3obMvf6Bdww9LOWdwOPFVVf96nz68cP7eR5EI6vwtHGmIDfu+7gWvbVUWbgJer6vklKWDUZ89fyy86v4QOAa8CLwB7u6Z9gs6VGE8Dl3W17wF+tQ2/jU5IzAD/DTh1RJ/ji8Afzmr7VWBPV52PttcTdA53jHrdfwl4HHis/QM5e3bdbfxyOleTfH9M6p6hc8z3kfY6fiXOWK7vXusP+CSdMAN4fdt2Z9q2/LYxWMfvonP48LGu9Xw58IfHt3PgxrZuH6VzIv+fj0HdPb/3WXWHzn8E9v22/U8u1fv7OApJkoeJJEmGgSQJw0CShGEgScIwkCRhGEiSMAwkScD/B26PZOEsJK2UAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "protein_data['score'].hist(bins=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that allseqs data is similar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "selection coeff cutoff = -0.815\n",
      "0.49173347278185797\n"
     ]
    }
   ],
   "source": [
    "# Classify the selection coeff series after fitting to a GMM\n",
    "gmm_model = TwoGaussianMixtureModelLabeler(\n",
    "    protein_data[protein_data.category.isin(['stop', 'wild_type'])].score)\n",
    "\n",
    "protein_data['viral_selection_gmm'] = gmm_model.predict(protein_data.score)\n",
    "\n",
    "# Compute the threshold for the viable class from the GMM labels\n",
    "selection_coeff_threshold = protein_data.loc[protein_data.viral_selection_gmm == 2, 'score'].min()\n",
    "print('selection coeff cutoff = %.3f' % selection_coeff_threshold)\n",
    "\n",
    "# Add a label column\n",
    "def is_viable_mutant(mutant_data):\n",
    "    return mutant_data['score'] > selection_coeff_threshold\n",
    "\n",
    "protein_data['binary_score'] = protein_data.apply(is_viable_mutant, axis=1)\n",
    "\n",
    "print(protein_data.binary_score.mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "design_train = [\n",
    "    'previous_chip_viable', 'previous_chip_nonviable', \n",
    "    'stop', 'singles', 'wild_type', 'random_doubles'\n",
    "]\n",
    "\n",
    "design_test = [\n",
    "    'rnn_designed_plus_rand_train_walked',\n",
    "    'rnn_designed_plus_rand_train_seed',\n",
    "    'rnn_rand_doubles_plus_singles_walked',\n",
    "    'rnn_rand_doubles_plus_singles_seed', 'rnn_standard_walked',\n",
    "    'rnn_standard_seed', 'cnn_designed_plus_rand_train_walked',\n",
    "    'cnn_designed_plus_rand_train_seed',\n",
    "    'cnn_rand_doubles_plus_single_walked',\n",
    "    'cnn_rand_doubles_plus_single_seed', 'cnn_standard_walked',\n",
    "    'cnn_standard_seed', 'lr_designed_plus_rand_train_walked',\n",
    "    'lr_designed_plus_rand_train_seed',\n",
    "    'lr_rand_doubles_plus_single_walked',\n",
    "    'lr_rand_doubles_plus_single_seed', 'lr_standard_walked',\n",
    "    'lr_standard_seed'\n",
    "]\n",
    "\n",
    "def design_task(category):\n",
    "    if category in design_train:\n",
    "        return \"train\"\n",
    "    elif category in design_test:\n",
    "        return \"test\"\n",
    "    else:\n",
    "        return \"unknown\"\n",
    "    \n",
    "def design_task_reversed(category):\n",
    "    if category in design_train:\n",
    "        return \"test\"\n",
    "    elif category in design_test:\n",
    "        return \"train\"\n",
    "    else:\n",
    "        return \"unknown\"\n",
    "\n",
    "protein_data['design_task'] = protein_data.category.apply(design_task)\n",
    "protein_data['design_task_reversed'] = protein_data.category.apply(design_task_reversed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The test set will be 75.2653179379672% of the data.\n"
     ]
    }
   ],
   "source": [
    "natural_allowed = [\n",
    "    'previous_chip_viable', 'previous_chip_nonviable', \n",
    "    'stop', 'singles', 'wild_type', 'random_doubles'\n",
    "]\n",
    "\n",
    "# randomly pick a few, see distribution of mutations\n",
    "def natural_task_1(category):\n",
    "    if category in natural_allowed:\n",
    "        return random.choices(['train', 'test'], weights=[0.8, 0.2], k=1)[0]\n",
    "    else:\n",
    "        return \"unknown\"\n",
    "\n",
    "# from train, those w/ many mutations\n",
    "\n",
    "num_mutations_threshold = 7\n",
    "\n",
    "test_count = len(protein_data[protein_data.number_of_mutations > num_mutations_threshold])\n",
    "total = len(protein_data)\n",
    "\n",
    "test_percentage = (test_count*100)/total\n",
    "\n",
    "print(f'The test set will be {test_percentage}% of the data.')\n",
    "\n",
    "def natural_task_2(row):\n",
    "    if row.category in natural_allowed:\n",
    "        if row.number_of_mutations > num_mutations_threshold:\n",
    "            return \"test\"\n",
    "        else:\n",
    "            return \"train\"\n",
    "    else:\n",
    "        return \"unknown\"\n",
    "    \n",
    "\n",
    "protein_data['natural_task_1'] = protein_data.category.apply(natural_task_1)\n",
    "protein_data['natural_task_2'] = protein_data.apply(natural_task_2, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Random sample n sequences from train k times for single shot learning approaches\n",
    "\n",
    "n = 96\n",
    "k = 100\n",
    "\n",
    "protein_data['natural_task_1_random_train_set'] = 'unassigned'\n",
    "protein_data['natural_task_2_random_train_set'] = 'unassigned'\n",
    "\n",
    "natural_task_1_train = protein_data.query('natural_task_1 == \"train\"')\n",
    "natural_task_1_train_index = list(natural_task_1_train.index)\n",
    "\n",
    "natural_task_1_kn_indexes = numpy.array(random.sample(natural_task_1_train_index, k=k*n)).reshape(k,n)\n",
    "\n",
    "for i, kn_index in enumerate(natural_task_1_kn_indexes):\n",
    "    protein_data.loc[kn_index, 'natural_task_1_random_train_set'] = f\"S{i+1}\"\n",
    "    \n",
    "natural_task_2_train = protein_data.query('natural_task_2 == \"train\"')\n",
    "natural_task_2_train_index = list(natural_task_2_train.index)\n",
    "\n",
    "natural_task_2_kn_indexes = numpy.array(random.sample(natural_task_2_train_index, k=k*n)).reshape(k,n)\n",
    "\n",
    "for i, kn_index in enumerate(natural_task_2_kn_indexes):\n",
    "    protein_data.loc[kn_index, 'natural_task_2_random_train_set'] = f\"S{i+1}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Set design_task has 28102 train and 186075 test sequences. 55% of test is positive.\n",
      "    \n",
      "\n",
      "    Set design_task_reversed has 186075 train and 28102 test sequences. 12% of test is positive.\n",
      "    \n",
      "\n",
      "    Set natural_task_1 has 22497 train and 5605 test sequences. 12% of test is positive.\n",
      "    \n",
      "\n",
      "    Set natural_task_2 has 27661 train and 441 test sequences. 13% of test is positive.\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "# Check how much are in test/train for each set\n",
    "\n",
    "def get_stats(task_name):\n",
    "    return f\"\"\"\n",
    "    Set {task_name} has {\n",
    "      len(protein_data.query(f\"{task_name}=='train'\"))\n",
    "    } train and {\n",
    "      len(protein_data.query(f\"{task_name}=='test'\"))\n",
    "    } test sequences. {\n",
    "      round(len(protein_data.query(f\"{task_name}=='test' & binary_score==True\")) / \n",
    "      len(protein_data.query(f\"{task_name}=='test'\")) * 100)\n",
    "    }% of test is positive.\n",
    "    \"\"\"\n",
    "\n",
    "print(get_stats(\"design_task\"))\n",
    "print(get_stats(\"design_task_reversed\"))\n",
    "print(get_stats(\"natural_task_1\"))\n",
    "print(get_stats(\"natural_task_2\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add reference region for completeness\n",
    "protein_data['reference_region'] = reference_region\n",
    "\n",
    "protein_data[\n",
    "    ['mutation_mask', 'mutated_region', 'category', 'number_of_mutations',\n",
    "    'levenshtein_distance', 'score', 'full_aa_sequence', 'binary_score', 'reference_region',\n",
    "    'design_task', 'design_task_reversed', 'natural_task_1', 'natural_task_2',\n",
    "    'natural_task_1_random_train_set', 'natural_task_2_random_train_set']\n",
    "].to_csv(task_path / \"full_data.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "ml_categories = ['train','test']\n",
    "\n",
    "# Design regression\n",
    "design_task_regression = protein_data[protein_data.design_task.isin(ml_categories)]\n",
    "design_task_regression = design_task_regression.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'score': 'target',\n",
    "    'design_task': 'set',\n",
    "}, axis=1)\n",
    "\n",
    "design_task_regression[\n",
    "    ['sequence', 'target', 'set']\n",
    "].to_csv(task_path / 'tasks' / \"design_task_regression.csv\", index=False)\n",
    "\n",
    "del design_task_regression\n",
    "\n",
    "# Design binary\n",
    "design_task_binary = protein_data[protein_data.design_task.isin(ml_categories)]\n",
    "design_task_binary = design_task_binary.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'binary_score': 'target',\n",
    "    'design_task': 'set',\n",
    "}, axis=1)\n",
    "\n",
    "design_task_binary[\n",
    "    ['sequence', 'target', 'set']\n",
    "].to_csv(task_path / 'tasks' / \"design_task_binary.csv\", index=False)\n",
    "\n",
    "del design_task_binary\n",
    "\n",
    "# Design reversed regression\n",
    "design_task_reversed_regression = protein_data[protein_data.design_task_reversed.isin(ml_categories)]\n",
    "design_task_reversed_regression = design_task_reversed_regression.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'score': 'target',\n",
    "    'design_task_reversed': 'set',\n",
    "}, axis=1)\n",
    "\n",
    "design_task_reversed_regression[\n",
    "    ['sequence', 'target', 'set']\n",
    "].to_csv(task_path / 'tasks' / \"design_task_reversed_regression.csv\", index=False)\n",
    "\n",
    "del design_task_reversed_regression\n",
    "\n",
    "# Design reversed binary\n",
    "design_task_reversed_binary = protein_data[protein_data.design_task_reversed.isin(ml_categories)]\n",
    "design_task_reversed_binary = design_task_reversed_binary.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'binary_score': 'target',\n",
    "    'design_task_reversed': 'set',\n",
    "}, axis=1)\n",
    "\n",
    "design_task_reversed_binary[\n",
    "    ['sequence', 'target', 'set']\n",
    "].to_csv(task_path / 'tasks' / \"design_task_reversed_binary.csv\", index=False)\n",
    "\n",
    "del design_task_reversed_binary\n",
    "\n",
    "# Natural 1 regression\n",
    "natural_task_1_regression = protein_data[protein_data.natural_task_1.isin(ml_categories)]\n",
    "natural_task_1_regression = natural_task_1_regression.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'score': 'target',\n",
    "    'natural_task_1': 'set',\n",
    "    'natural_task_1_random_train_set': 'random_sample'\n",
    "}, axis=1)\n",
    "\n",
    "natural_task_1_regression[\n",
    "    ['sequence', 'target', 'set', 'random_sample']\n",
    "].to_csv(task_path / 'tasks' / \"natural_task_1_regression.csv\", index=False)\n",
    "\n",
    "del natural_task_1_regression\n",
    "\n",
    "# Natural 1 binary\n",
    "natural_task_1_binary = protein_data[protein_data.natural_task_1.isin(ml_categories)]\n",
    "natural_task_1_binary = natural_task_1_binary.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'binary_score': 'target',\n",
    "    'natural_task_1': 'set',\n",
    "    'natural_task_1_random_train_set': 'random_sample'\n",
    "}, axis=1)\n",
    "\n",
    "natural_task_1_binary[\n",
    "    ['sequence', 'target', 'set', 'random_sample']\n",
    "].to_csv(task_path / 'tasks' / \"natural_task_1_binary.csv\", index=False)\n",
    "\n",
    "del natural_task_1_binary\n",
    "\n",
    "# Natural 2 regression\n",
    "natural_task_2_regression = protein_data[protein_data.natural_task_2.isin(ml_categories)]\n",
    "natural_task_2_regression = natural_task_2_regression.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'score': 'target',\n",
    "    'natural_task_2': 'set',\n",
    "    'natural_task_2_random_train_set': 'random_sample'\n",
    "}, axis=1)\n",
    "\n",
    "natural_task_2_regression[\n",
    "    ['sequence', 'target', 'set', 'random_sample']\n",
    "].to_csv(task_path / 'tasks' / \"natural_task_2_regression.csv\", index=False)\n",
    "\n",
    "del natural_task_2_regression\n",
    "\n",
    "# Natural 2 binary\n",
    "natural_task_2_binary = protein_data[protein_data.natural_task_2.isin(ml_categories)]\n",
    "natural_task_2_binary = natural_task_2_binary.rename({\n",
    "    'full_aa_sequence': 'sequence',\n",
    "    'binary_score': 'target',\n",
    "    'natural_task_2': 'set',\n",
    "    'natural_task_2_random_train_set': 'random_sample'\n",
    "}, axis=1)\n",
    "\n",
    "natural_task_2_binary[\n",
    "    ['sequence', 'target', 'set', 'random_sample']\n",
    "].to_csv(task_path / 'tasks' / \"natural_task_2_binary.csv\", index=False)\n",
    "\n",
    "del natural_task_2_binary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
